In [ ]:
# import numerical python
import numpy as np

# import the pandas
import pandas as pd

#import the seaborn
import seaborn as sns

#import the matplotlib
import matplotlib.pyplot as plt

In [ ]:
# load the data
df_House_Price1 = pd.read_csv('House_Price1.csv')

In [ ]:
# show the first five observations of the data
df_House_Price1.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter  Sold  
0     YES       5.480      11.1920     River        23     YES     0  
1      NO       7.332      12.1728      Lake        42     YES     1  
2      NO       7.394     101.1200       NaN        38     YES     0  
3     YES       9.268      11.2672      Lake        45     YES     0  
4      NO       8.824      11.2896      Lake        55     YES     0

In [ ]:
# show the last five observations of the data
df_House_Price1.tail()

price  resid_area  air_qual  room_num   age  distance  teachers  \
501   22.4       41.93     0.573     6.593  69.1      2.64      19.0   
502   20.6       41.93     0.573     6.120  76.7      2.44      19.0   
503   23.9       41.93     0.573     6.976  91.0      2.34      19.0   
504   22.0       41.93     0.573     6.794  89.3      2.54      19.0   
505   19.0       41.93     0.573     6.030  80.8      2.72      19.0   

     poor_prop airport  n_hos_beds  n_hot_rooms       waterbody  rainfall  \
501       9.67      NO       9.348      12.1792  Lake and River        27   
502       9.08     YES       6.612      13.1648  Lake and River        20   
503       5.64      NO       5.478      12.1912             NaN        31   
504       6.48     YES       7.940      15.1760             NaN        47   
505       7.88     YES      10.280      10.1520             NaN        45   

    bus_ter  Sold  
501     YES     1  
502     YES     1  
503     YES     1  
504     YES     1  
505     YES     1

In [ ]:
# show the shape of the data
df_House_Price1.shape

(506, 15)

In [ ]:
# show the basic staistics of the data
df_House_Price1.describe()

price  resid_area    air_qual    room_num         age    distance  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    22.528854   41.136779    0.554695    6.284634   68.574901    3.971996   
std      9.182176    6.860353    0.115878    0.702617   28.148861    2.108532   
min      5.000000   30.460000    0.385000    3.561000    2.900000    1.130000   
25%     17.025000   35.190000    0.449000    5.885500   45.025000    2.270000   
50%     21.200000   39.690000    0.538000    6.208500   77.500000    3.385000   
75%     25.000000   48.100000    0.624000    6.623500   94.075000    5.367500   
max     50.000000   57.740000    0.871000    8.780000  100.000000   12.320000   

         teachers   poor_prop  n_hos_beds  n_hot_rooms    rainfall        Sold  
count  506.000000  506.000000  498.000000   506.000000  506.000000  506.000000  
mean    21.544466   12.653063    7.899767    13.041605   39.181818    0.454545  
std      2.164946    7.141062    1.476683     5.238957   12.513697    0.498422  
min     18.000000    1.730000    5.268000    10.057600    3.000000    0.000000  
25%     19.800000    6.950000    6.634500    11.189800   28.000000    0.000000  
50%     20.950000   11.360000    7.999000    12.720000   39.000000    0.000000  
75%     22.600000   16.955000    9.088000    14.170800   50.000000    1.000000  
max     27.400000   37.970000   10.876000   101.120000   60.000000    1.000000

In [ ]:
# show the basic informations of the data
df_House_Price1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  Sold         506 non-null    int64  
dtypes: float64(10), int64(2), object(3)
memory usage: 59.4+ KB


Interpretation

 - we have 506 observations and 15 attributes
 - we have age column which contain the outliers


**Data Pre-Processing**

In [ ]:
# Lets perform the data cleaning
df_House_Price1.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'Sold'],
      dtype='object')

In [ ]:
dictionary={
    'price':'Price',
    'resid_area':'Residential Area',
    'air_quality':'Air Quality',
    'room_number':'Room Number',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Poor_Property',
    'airport':'Airport',
    'n_hos_beds':'N_Hospital_Beds',
    'n_hot_rooms':'N_Hotet_Rooms',
    'waterbody':'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Ter',
    'sold':'Sold'
}

In [ ]:
# perform the rename action
df_House_Price1.rename(columns=dictionary)

Price  Residential Area  air_qual  room_num   Age  Distance  Teachers  \
0     24.0             32.31     0.538     6.575  65.2      4.35      24.7   
1     21.6             37.07     0.469     6.421  78.9      4.99      22.2   
2     34.7             37.07     0.469     7.185  61.1      5.03      22.2   
3     33.4             32.18     0.458     6.998  45.8      6.21      21.3   
4     36.2             32.18     0.458     7.147  54.2      6.16      21.3   
..     ...               ...       ...       ...   ...       ...       ...   
501   22.4             41.93     0.573     6.593  69.1      2.64      19.0   
502   20.6             41.93     0.573     6.120  76.7      2.44      19.0   
503   23.9             41.93     0.573     6.976  91.0      2.34      19.0   
504   22.0             41.93     0.573     6.794  89.3      2.54      19.0   
505   19.0             41.93     0.573     6.030  80.8      2.72      19.0   

     Poor_Property Airport  N_Hospital_Beds  N_Hotet_Rooms       Waterbody  \
0             4.98     YES            5.480        11.1920           River   
1             9.14      NO            7.332        12.1728            Lake   
2             4.03      NO            7.394       101.1200             NaN   
3             2.94     YES            9.268        11.2672            Lake   
4             5.33      NO            8.824        11.2896            Lake   
..             ...     ...              ...            ...             ...   
501           9.67      NO            9.348        12.1792  Lake and River   
502           9.08     YES            6.612        13.1648  Lake and River   
503           5.64      NO            5.478        12.1912             NaN   
504           6.48     YES            7.940        15.1760             NaN   
505           7.88     YES           10.280        10.1520             NaN   

     Rainfall Bus_Ter  Sold  
0          23     YES     0  
1          42     YES     1  
2          38     YES     0  
3          45     YES     0  
4          55     YES     0  
..        ...     ...   ...  
501        27     YES     1  
502        20     YES     1  
503        31     YES     1  
504        47     YES     1  
505        45     YES     1  

[506 rows x 15 columns]

In [ ]:
# Do the authentic check
df_House_Price1.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'Sold'],
      dtype='object')

**Null Value Handling**

In [ ]:
# Find the number of missing records
df_House_Price1.isnull().sum()

price            0
resid_area       0
air_qual         0
room_num         0
age              0
distance         0
teachers         0
poor_prop        0
airport          0
n_hos_beds       8
n_hot_rooms      0
waterbody      155
rainfall         0
bus_ter          0
Sold             0
dtype: int64

In [ ]:
# Find the percentage
df_House_Price1.isnull().sum()/df_House_Price1.shape[0]*100

price           0.000000
resid_area      0.000000
air_qual        0.000000
room_num        0.000000
age             0.000000
distance        0.000000
teachers        0.000000
poor_prop       0.000000
airport         0.000000
n_hos_beds      1.581028
n_hot_rooms     0.000000
waterbody      30.632411
rainfall        0.000000
bus_ter         0.000000
Sold            0.000000
dtype: float64

In [ ]:
# find the mode of waterbody column
df_House_Price1.waterbody.mode()

0    River
Name: waterbody, dtype: object

In [ ]:
# Lets treat the null values in the waterbody column
df_House_Price1.waterbody=df_House_Price1.waterbody.fillna('River')

In [ ]:
# Do the authentic check
df_House_Price1.isnull().sum()

price          0
resid_area     0
air_qual       0
room_num       0
age            0
distance       0
teachers       0
poor_prop      0
airport        0
n_hos_beds     8
n_hot_rooms    0
waterbody      0
rainfall       0
bus_ter        0
Sold           0
dtype: int64

In [ ]:
# In N_Hospital_Beds, we have less than 10% of data is missing
df_House_Price1.dropna(inplace=True)

In [ ]:
df_target=df_House_Price1['price']

In [ ]:
df_features=df_House_Price1.drop('price',axis=1)

In [ ]:
# Segregate the data based on the datatype
df_num=df_House_Price1.select_dtypes(include='number')
df_cat=df_House_Price1.select_dtypes(include='object')

In [ ]:
# show the numerical data
df_num.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

   n_hos_beds  n_hot_rooms  rainfall  Sold  
0       5.480      11.1920        23     0  
1       7.332      12.1728        42     1  
2       7.394     101.1200        38     0  
3       9.268      11.2672        45     0  
4       8.824      11.2896        55     0

In [ ]:
df_cat.head()

airport waterbody bus_ter
0     YES     River     YES
1      NO      Lake     YES
2      NO     River     YES
3     YES      Lake     YES
4      NO      Lake     YES

**Perform Encoding on categorical variables**

In [ ]:
# Use Dummy Encoding
df_encoded=pd.get_dummies(df_cat)
# Reset the index
df_encoded=df_encoded.reset_index()

In [ ]:
# show the encoded data
df_encoded.head()

index  airport_NO  airport_YES  waterbody_Lake  waterbody_Lake and River  \
0      0       False         True           False                     False   
1      1        True        False            True                     False   
2      2        True        False           False                     False   
3      3       False         True            True                     False   
4      4        True        False            True                     False   

   waterbody_River  bus_ter_YES  
0             True         True  
1            False         True  
2             True         True  
3            False         True  
4            False         True

**Perform Scalling On Numerical Data**

In [ ]:
# import the standard scaler
from sklearn.preprocessing import StandardScaler

# Initiate the standard scaler
ss=StandardScaler()

# scale the value present in numerical dataframe
scaled=ss.fit_transform(df_num)

# Numerate the data into the dataframe
scaled_df=pd.DataFrame(scaled,columns=df_num.columns)

# Reset the index
scaled_df=scaled_df.reset_index()

In [ ]:
# show the data
scaled_df.head()

index     price  resid_area  air_qual  room_num       age  distance  \
0      0  0.154792   -1.282444 -0.137094  0.408975 -0.115644  0.172845   
1      1 -0.106592   -0.590059 -0.733195  0.189974  0.369741  0.476100   
2      2  1.320127   -0.590059 -0.733195  1.276446 -0.260905  0.495053   
3      3  1.178544   -1.301354 -0.828226  1.010516 -0.802977  1.054179   
4      4  1.483492   -1.301354 -0.828226  1.222407 -0.505369  1.030487   

   teachers  poor_prop  n_hos_beds  n_hot_rooms  rainfall      Sold  
0  1.462167  -1.066368   -1.640299    -0.352111 -1.302068 -0.911527  
1  0.304963  -0.484136   -0.384875    -0.166019  0.215194  1.097060  
2  0.304963  -1.199330   -0.342847    16.710340 -0.104230 -0.911527  
3 -0.111631  -1.351886    0.927490    -0.337843  0.454762 -0.911527  
4 -0.111631  -1.017382    0.626514    -0.333593  1.253320 -0.911527

**Perform the Concatenate Operation**

In [ ]:
# Concate encoded variable and scaled variable
df_independent=pd.concat([df_encoded,scaled_df],axis=1)

In [ ]:
# show the data
df_independent.head()

index  airport_NO  airport_YES  waterbody_Lake  waterbody_Lake and River  \
0      0       False         True           False                     False   
1      1        True        False            True                     False   
2      2        True        False           False                     False   
3      3       False         True            True                     False   
4      4        True        False            True                     False   

   waterbody_River  bus_ter_YES  index     price  resid_area  air_qual  \
0             True         True      0  0.154792   -1.282444 -0.137094   
1            False         True      1 -0.106592   -0.590059 -0.733195   
2             True         True      2  1.320127   -0.590059 -0.733195   
3            False         True      3  1.178544   -1.301354 -0.828226   
4            False         True      4  1.483492   -1.301354 -0.828226   

   room_num       age  distance  teachers  poor_prop  n_hos_beds  n_hot_rooms  \
0  0.408975 -0.115644  0.172845  1.462167  -1.066368   -1.640299    -0.352111   
1  0.189974  0.369741  0.476100  0.304963  -0.484136   -0.384875    -0.166019   
2  1.276446 -0.260905  0.495053  0.304963  -1.199330   -0.342847    16.710340   
3  1.010516 -0.802977  1.054179 -0.111631  -1.351886    0.927490    -0.337843   
4  1.222407 -0.505369  1.030487 -0.111631  -1.017382    0.626514    -0.333593   

   rainfall      Sold  
0 -1.302068 -0.911527  
1  0.215194  1.097060  
2 -0.104230 -0.911527  
3  0.454762 -0.911527  
4  1.253320 -0.911527

**Split the Data for Training and Testing**

In [ ]:
# import the train test split
from sklearn.model_selection import train_test_split


In [ ]:
# split the data
x_train,x_test,y_train,y_test=train_test_split(df_independent,
                                               df_target,
                                               test_size=0.2)

In [ ]:
# check the size
x_train.shape

(398, 20)

In [ ]:
x_test.shape

(100, 20)

In [ ]:
y_train.shape

(398,)

In [ ]:
y_test.shape

(100,)

**Apply Linear Regression**

In [ ]:
# import the model
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [ ]:
# initiate the model
LR=LinearRegression()

In [ ]:
# Train the linear regression model
model_lr=LR.fit(x_train,y_train)

In [ ]:
# perform the prediction
predicted=model_lr.predict(x_test)

In [ ]:
# show the predicted value
predicted

array([36.4, 22.7, 22.6, 10.4,  8.1, 11.5, 30.1, 14.1, 22.4, 21.9,  5. ,
       13.8, 15.4, 18.5, 22.8, 23.8, 14.2, 22.2, 19.2, 18.9, 10.2, 46. ,
       26.2, 22.2, 15.4, 19.1, 20.4, 13.4, 12.1, 21.2, 24.1, 20.1, 18.9,
       50. , 33.4, 50. , 22. , 19.9, 31.5, 19.6, 17.9, 18.4, 23.8, 20.6,
       48.8, 20.6, 15.2, 10.5, 18.7, 24.5, 20.6, 22.2, 50. , 33.4, 17.2,
       24.8, 19.3, 11.8, 30.5, 22. , 33. , 14.1, 19.1, 19.5, 23.7, 27.5,
       22.3, 21.7, 21. , 23.3, 36.5, 38.7, 22.2, 21.7, 19.5, 19.6, 30.1,
       22.5, 21.7, 34.9, 13.8, 19.5, 27. , 17.8, 24.4, 18.5, 22.9, 13.8,
       20.1, 17. ,  5. , 19.3, 28.7, 23.8, 20.2, 19. , 17.6,  9.7, 22. ,
       19.4])

In [ ]:
# Evaluate the model
round(mean_squared_error(predicted,y_test),3)

0.0

**Lab 1- First neural network with keras**

In [ ]:
from numpy import loadtxt
from tensorflow import keras

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
